app/scripts/lib/createDupeReqFilterMiddleware.js

```diff
@@ -0,0 +1,23 @@
+import log from 'loglevel';
+
+/**
+ * Returns a middleware that filters out requests already seen
+ *
+ * @returns {Function}
+ */
+export default function createDupeReqFilterMiddleware() {
+  const processedRequestId = [];
+  return function filterDuplicateRequestMiddleware(
+    /** @type {any} */ req,
+    /** @type {any} */ _,
+    /** @type {Function} */ next,
+    /** @type {Function} */ end,
+  ) {
+    if (processedRequestId.indexOf(req.id) >= 0) {
+      log.info(`RPC request with id ${req.id} already seen.`);
+      return end();
```

Comment:
Are we sure the dapp has received the response already from the first request? i.e. is it possible for the first response to have been dropped, e.g. while the contentscript was waiting for the initialization message.

Reply:
I suspect it is possible, though maybe not likely. It could happen after the contentscript has destroyed the streams, before they had been recreated.

It might be safer to cache the response, and re-send the original response if we get the message again. Not great having to cache every dapp response though :thinking: Even for just 5 minutes, that could get hefty.

Reply:
Alternatively, we could add a retry loop on the service worker side to ensure the request gets through if this unfortunate timing happens. Then this middleware works fine as-is.

Reply:
The filters are called after request reaches service worker, the only scenario I think when user may not get response is when service worker dies before request is completely processed, in that case replay logic in `inpage` will resend. the request.

Reply:
Here is the failure scenario I was thinking of: https://sequencediagram.org/index.html#initialData=C4S2BsFMAIDEEMTgK4CcYGcDGkB29UQB7AKBIAcDQsRLdhoBhI+vYbQ84CqkGuhgGVIqAG58YAdyKoA1iOgAKAIwZgASh6pqtePWjCxE6NLkLFAJlwATTSXhZQo+MBjNW9DrW4OnLmIbiOCYy8qhKqhpk7q6eWJzAALQAfIHGpmERauoAXAAGkMAAFgD6WEWIuACS1nnQ6ACOyJBqZNaQviDOrgYiQVKh5pF2ZPAARkQAHvUgAOZFDEQAZr1GwRlD2TkACqhEOBgY0Hjt1qMT04Tziytp64Phlja5u-stR2pUkGed3QF96QeSistmiLFi7Hi3hSdwGZkeINyBWKJQwJwAKqg9BhOiw6o1mq0YmwvFwADyJWEheFZDT5QqlcqVGp1NToeAAW2gS0QKHQZHsF2gUCWNyY4JJUK4OQAIi1gHsAJ7HSaxDDEXDQNkdDkYEjEuIJClUjYI570lFomyY7G43B1SL1FrkFho85TYWQUXQZbijyQhI5ABKkHiHR6kFVeHVLC1Cp1ev1EsN0NSAPuNKetgtpSt1htuBxjg1+MgTXlYP9pOAxvTcMyWaRDNRGKxhbtdUgHK4ypBTowLsLkCAA

I'm not sure whether it's possible. The main part I'm uncertain of is whether messages sent over the "contentscript <=> service worker" stream for the _old_ process would be delivered to the _new_ process. Maybe the stream needs to be re-established.

Reply:
Not sure if I am getting you correct, but in case it is not delivered to backend - there will be no response and DAPP replay logic will replay it.

Reply:
Got it, thanks. That sounds right. I just wasn't sure.